### ISSUE TITLE
websocket 0-rtt 配置无效

### ISSUE THREAD

**Issue by escapezn** [MetaCubeX/Clash.Meta#224]
服务端是xray1.6.0，clash配置了ws early data不能起到减少rtt的作用。使用xray作为客户端连接，带有?ed=2048参数的配置测试延迟比不带的少1个rtt
![1](https://user-images.githubusercontent.com/30799958/212556721-0f77c0e8-96e9-415d-aef2-98c079138004.png)

clash配置
```
  - name: "ssgr"
    type: ss
    server: 
    port: 80
    cipher: aes-128-gcm
    password: 
    udp: true
    udp-over-tcp: true
    plugin: v2ray-plugin
    plugin-opts:
      mode: websocket # no QUIC now
      tls: false
      host: 
      path: "/path?ed=2048"
      mux: false
      # headers:
      #   custom: value

  - name: "socks"
    type: socks5
    server: 127.0.0.1
    port: 2080
    udp: true
```
xray配置
![2](https://user-images.githubusercontent.com/30799958/212556714-c00f3651-ae27-4565-8e04-b0a6bf7b1dc1.png)

**Comment by malikshi** [comment]
Did you try with headers options? Is work?
    headers:
      Host: hostname
      max-early-data: 2048
      early-data-header-name: Sec-WebSocket-Protocol
  udp: true

**Comment by escapezn** [comment]
> Did you try with headers options? Is work?
> 
> ```
>     headers:
>       Host: hostname
>       max-early-data: 2048
>       early-data-header-name: Sec-WebSocket-Protocol
> ```


原版clash文档记载的两种写法，一开始就写的上面这种，都没有用
```
proxies:
# for v2fly and xray (recommand)
  - name: "vmess-ws-0rtt"
    type: vmess
    server: server
    port: 443
    uuid: uuid
    alterId: 32
    cipher: auto
    network: ws
    ws-opts:
      path: /
      max-early-data: 2048
      early-data-header-name: Sec-WebSocket-Protocol
# for xray compatible (not recommended)
  - name: "vmess-ws-0rtt"
    type: vmess
    server: server
    port: 443
    uuid: uuid
    alterId: 32
    cipher: auto
    network: ws
    ## also work for xray
    # ws-opts:
    #  path: /
    #  max-early-data: 2048
    #  early-data-header-name: Sec-WebSocket-Protocol
    ws-opts:
      path: /?ed=2048
```

**Comment by wwqgtxx** [comment]
你有测试过原版的clash核心延迟是否有区别么

**Comment by escapezn** [comment]
> 你有测试过原版的clash核心延迟是否有区别么

和meta是一样的

**Comment by moranno** [comment]
我这边测试原版clash和clash meta均无效。
尝试了2种配置文件写法均无效：
```
- { name: (vmess), type: vmess, server: x.x.x port: 1234, uuid: xxx, alterId: 0, cipher: auto, udp: true, network: ws, ws-opts: { ws-path: '/', max-early-data: 2048, early-data-header-name: Sec-WebSocket-Protocol } }
- { name: (vmess), type: vmess, server: x.x.x port: 1234, uuid: xxx, alterId: 0, cipher: auto, udp: true, network: ws, ws-opts: { ws-path: '/?ed=2048 } }
```

相关帖子：
https://github.com/Dreamacro/clash/issues/2400
https://github.com/Dreamacro/clash/issues/2531
https://github.com/Dreamacro/clash/pull/1558
https://github.com/SagerNet/sing-box/issues/169

**Comment by wwqgtxx** [comment]
fixed in https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa
目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效

**Comment by sqwwqw5** [comment]
> fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效

Why isn't useEarly true for shadowsocks-WebSocket?
It should be true.

**Comment by wwqgtxx** [comment]
> > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> 
> Why isn't useEarly true for shadowsocks-WebSocket? It should be true.

请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的

**Comment by sqwwqw5** [comment]
> > > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> > 
> > 
> > Why isn't useEarly true for shadowsocks-WebSocket? It should be true.
> 
> 请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的

	case "websocket":
		var err error
		c, err = v2rayObfs.NewV2rayObfs(c, ss.v2rayOption)
		if err != nil {
			return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
		}
	case shadowtls.Mode:
		var err error
		c, err = shadowtls.NewShadowTLS(ctx, c, ss.shadowTLSOption)
		if err != nil {
			return nil, err
		}
		useEarly = true
	case restls.Mode:
		var err error
		c, err = restls.NewRestls(ctx, c, ss.restlsConfig)
		if err != nil {
			return nil, fmt.Errorf("%s (restls) connect error: %w", ss.addr, err)
		}
		useEarly = true
	}

See the difference? It doesn't return useEarly = true for case "websocket" and by testing on clash.meta client/xray server it doesn't work either.

**Comment by wwqgtxx** [comment]
> > > > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> > > 
> > > 
> > > Why isn't useEarly true for shadowsocks-WebSocket? It should be true.
> > 
> > 
> > 请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的
> 
> ```
> case "websocket":
> 	var err error
> 	c, err = v2rayObfs.NewV2rayObfs(c, ss.v2rayOption)
> 	if err != nil {
> 		return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
> 	}
> case shadowtls.Mode:
> 	var err error
> 	c, err = shadowtls.NewShadowTLS(ctx, c, ss.shadowTLSOption)
> 	if err != nil {
> 		return nil, err
> 	}
> 	useEarly = true
> case restls.Mode:
> 	var err error
> 	c, err = restls.NewRestls(ctx, c, ss.restlsConfig)
> 	if err != nil {
> 		return nil, fmt.Errorf("%s (restls) connect error: %w", ss.addr, err)
> 	}
> 	useEarly = true
> }
> ```
> 
> See the difference? It doesn't return useEarly = true for case "websocket" and by testing on clash.meta client/xray server it doesn't work either.

你就不能在往下看一行么
```golang
useEarly = useEarly || N.NeedHandshake(c)
```

**Comment by sqwwqw5** [comment]
> > > > > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> > > > 
> > > > 
> > > > Why isn't useEarly true for shadowsocks-WebSocket? It should be true.
> > > 
> > > 
> > > 请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的
> > 
> > 
> > ```
> > case "websocket":
> > 	var err error
> > 	c, err = v2rayObfs.NewV2rayObfs(c, ss.v2rayOption)
> > 	if err != nil {
> > 		return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
> > 	}
> > case shadowtls.Mode:
> > 	var err error
> > 	c, err = shadowtls.NewShadowTLS(ctx, c, ss.shadowTLSOption)
> > 	if err != nil {
> > 		return nil, err
> > 	}
> > 	useEarly = true
> > case restls.Mode:
> > 	var err error
> > 	c, err = restls.NewRestls(ctx, c, ss.restlsConfig)
> > 	if err != nil {
> > 		return nil, fmt.Errorf("%s (restls) connect error: %w", ss.addr, err)
> > 	}
> > 	useEarly = true
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > See the difference? It doesn't return useEarly = true for case "websocket" and by testing on clash.meta client/xray server it doesn't work either.
> 
> 你就不能在往下看一行么
> 
> ```go
> useEarly = useEarly || N.NeedHandshake(c)
> ```

My bad, I think I figured out the issue, the NeedHandshake fuction isn't defined in /transport/v2ray-plugin/websocket.go.

**Comment by wwqgtxx** [comment]
> > > > > > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> > > > > 
> > > > > 
> > > > > Why isn't useEarly true for shadowsocks-WebSocket? It should be true.
> > > > 
> > > > 
> > > > 请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的
> > > 
> > > 
> > > ```
> > > case "websocket":
> > > 	var err error
> > > 	c, err = v2rayObfs.NewV2rayObfs(c, ss.v2rayOption)
> > > 	if err != nil {
> > > 		return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
> > > 	}
> > > case shadowtls.Mode:
> > > 	var err error
> > > 	c, err = shadowtls.NewShadowTLS(ctx, c, ss.shadowTLSOption)
> > > 	if err != nil {
> > > 		return nil, err
> > > 	}
> > > 	useEarly = true
> > > case restls.Mode:
> > > 	var err error
> > > 	c, err = restls.NewRestls(ctx, c, ss.restlsConfig)
> > > 	if err != nil {
> > > 		return nil, fmt.Errorf("%s (restls) connect error: %w", ss.addr, err)
> > > 	}
> > > 	useEarly = true
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > See the difference? It doesn't return useEarly = true for case "websocket" and by testing on clash.meta client/xray server it doesn't work either.
> > 
> > 
> > 你就不能在往下看一行么
> > ```go
> > useEarly = useEarly || N.NeedHandshake(c)
> > ```
> 
> My bad, I think I figured out the issue, the NeedHandshake function isn't defined in /transport/v2ray-plugin/websocket.go.

如果你不能确定问题在哪，甚至是有没有问题请不要浪费大家的时间
`NewV2rayObfs`中的返回值是由`vmess.StreamWebsocketConn`产生的，`StreamWebsocketConn`会在有`earlyData`配置的时候返回`*websocketWithEarlyDataConn`，该类型是实现了`NeedHandshake`的

**Comment by sqwwqw5** [comment]
> > > > > > > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> > > > > > 
> > > > > > 
> > > > > > Why isn't useEarly true for shadowsocks-WebSocket? It should be true.
> > > > > 
> > > > > 
> > > > > 请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的
> > > > 
> > > > 
> > > > ```
> > > > case "websocket":
> > > > 	var err error
> > > > 	c, err = v2rayObfs.NewV2rayObfs(c, ss.v2rayOption)
> > > > 	if err != nil {
> > > > 		return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
> > > > 	}
> > > > case shadowtls.Mode:
> > > > 	var err error
> > > > 	c, err = shadowtls.NewShadowTLS(ctx, c, ss.shadowTLSOption)
> > > > 	if err != nil {
> > > > 		return nil, err
> > > > 	}
> > > > 	useEarly = true
> > > > case restls.Mode:
> > > > 	var err error
> > > > 	c, err = restls.NewRestls(ctx, c, ss.restlsConfig)
> > > > 	if err != nil {
> > > > 		return nil, fmt.Errorf("%s (restls) connect error: %w", ss.addr, err)
> > > > 	}
> > > > 	useEarly = true
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > See the difference? It doesn't return useEarly = true for case "websocket" and by testing on clash.meta client/xray server it doesn't work either.
> > > 
> > > 
> > > 你就不能在往下看一行么
> > > ```go
> > > useEarly = useEarly || N.NeedHandshake(c)
> > > ```
> > 
> > 
> > My bad, I think I figured out the issue, the NeedHandshake function isn't defined in /transport/v2ray-plugin/websocket.go.
> 
> 如果你不能确定问题在哪，甚至是有没有问题请不要浪费大家的时间 `NewV2rayObfs`中的返回值是由`vmess.StreamWebsocketConn`产生的，`StreamWebsocketConn`会在有`earlyData`配置的时候返回`*websocketWithEarlyDataConn`，该类型是实现了`NeedHandshake`的

I will admit I'm not fimilar with codes, but I do have problems with 0-RTT.
![image](https://user-images.githubusercontent.com/3581246/230805760-d0ee79c6-ac59-433d-b88d-a8e89cb02a8e.png)
Setup:
  - name: Korea-cdn
    type: ss
    server: ****
    port: "443"
    cipher: chacha20-ietf-poly1305
    password: "****"
    udp: false
    plugin: v2ray-plugin
    plugin-opts:
      mux: 0
      mode: websocket
      tls: true
      headers:
        Host: ****
  - name: Korea-cdn-ed
    type: ss
    server: ****
    port: "443"
    cipher: chacha20-ietf-poly1305
    password: "****"
    udp: false
    plugin: v2ray-plugin
    plugin-opts:
      mux: 0
      mode: websocket
      tls: true
      headers:
        Host: ****
        max-early-data: 2048
        early-data-header-name: Sec-WebSocket-Protocol

**Comment by wwqgtxx** [comment]
> > > > > > > > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> > > > > > > 
> > > > > > > 
> > > > > > > Why isn't useEarly true for shadowsocks-WebSocket? It should be true.
> > > > > > 
> > > > > > 
> > > > > > 请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的
> > > > > 
> > > > > 
> > > > > ```
> > > > > case "websocket":
> > > > > 	var err error
> > > > > 	c, err = v2rayObfs.NewV2rayObfs(c, ss.v2rayOption)
> > > > > 	if err != nil {
> > > > > 		return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
> > > > > 	}
> > > > > case shadowtls.Mode:
> > > > > 	var err error
> > > > > 	c, err = shadowtls.NewShadowTLS(ctx, c, ss.shadowTLSOption)
> > > > > 	if err != nil {
> > > > > 		return nil, err
> > > > > 	}
> > > > > 	useEarly = true
> > > > > case restls.Mode:
> > > > > 	var err error
> > > > > 	c, err = restls.NewRestls(ctx, c, ss.restlsConfig)
> > > > > 	if err != nil {
> > > > > 		return nil, fmt.Errorf("%s (restls) connect error: %w", ss.addr, err)
> > > > > 	}
> > > > > 	useEarly = true
> > > > > }
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > See the difference? It doesn't return useEarly = true for case "websocket" and by testing on clash.meta client/xray server it doesn't work either.
> > > > 
> > > > 
> > > > 你就不能在往下看一行么
> > > > ```go
> > > > useEarly = useEarly || N.NeedHandshake(c)
> > > > ```
> > > 
> > > 
> > > My bad, I think I figured out the issue, the NeedHandshake function isn't defined in /transport/v2ray-plugin/websocket.go.
> > 
> > 
> > 如果你不能确定问题在哪，甚至是有没有问题请不要浪费大家的时间 `NewV2rayObfs`中的返回值是由`vmess.StreamWebsocketConn`产生的，`StreamWebsocketConn`会在有`earlyData`配置的时候返回`*websocketWithEarlyDataConn`，该类型是实现了`NeedHandshake`的
> 
> I will admit I'm not fimilar with codes, but I do have problems with 0-RTT. ![image](https://user-images.githubusercontent.com/3581246/230805760-d0ee79c6-ac59-433d-b88d-a8e89cb02a8e.png) Setup:
> 
> * name: Korea-cdn
>   type: ss
>   server: ****
>   port: "443"
>   cipher: chacha20-ietf-poly1305
>   password: "****"
>   udp: false
>   plugin: v2ray-plugin
>   plugin-opts:
>   mux: 0
>   mode: websocket
>   tls: true
>   headers:
>   Host: ****
> * name: Korea-cdn-ed
>   type: ss
>   server: ****
>   port: "443"
>   cipher: chacha20-ietf-poly1305
>   password: "****"
>   udp: false
>   plugin: v2ray-plugin
>   plugin-opts:
>   mux: 0
>   mode: websocket
>   tls: true
>   headers:
>   Host: ****
>   max-early-data: 2048
>   early-data-header-name: Sec-WebSocket-Protocol

v2ray-plugin不支持`max-early-data`配置项，请在`path`中使用`?ed=2048`格式进行配置（同xray）

**Comment by sqwwqw5** [comment]
> * early-data-header-name: Sec-WebSocket-Protocol

I tested again and again, with the path scheme alone or incorporating the "early-data-header-name: Sec-WebSocket-Protocol" header.

To make sure the server setup is a go for 0-RTT I also tested with the Xray client and Xray-plugin client which does work and cut more than half a second of connection latency. 

I do not intend to waste your time, I will inspect the code further to see if there's something wrong or with my setup. 

Also, the path scheme might not work with the current v2ray-plugin/websocket.go, see the fix I committed to the Xray-plugin yesterday, this fix does work, not sure if it's relatable though. https://github.com/teddysun/xray-plugin/commit/cc2df0f46bf7607bad159878497b19acf83bc6a6

**Comment by sqwwqw5** [comment]
> > > > > > > fixed in [75680c5](https://github.com/MetaCubeX/Clash.Meta/commit/75680c5866ebc497016120cda9afc39c3e13a0aa) 目前仅对shadowsocks, vmess, vless生效，trojan的wss模式下目前无效
> > > > > > 
> > > > > > 
> > > > > > Why isn't useEarly true for shadowsocks-WebSocket? It should be true.
> > > > > 
> > > > > 
> > > > > 请仔细看目前shadowsocks.go中StreamConnContext的实现，N.NeedHandshake会判断v2rayObfs返回的net.Conn是否为0-rtt的ws连接的
> > > > 
> > > > 
> > > > ```
> > > > case "websocket":
> > > > 	var err error
> > > > 	c, err = v2rayObfs.NewV2rayObfs(c, ss.v2rayOption)
> > > > 	if err != nil {
> > > > 		return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
> > > > 	}
> > > > case shadowtls.Mode:
> > > > 	var err error
> > > > 	c, err = shadowtls.NewShadowTLS(ctx, c, ss.shadowTLSOption)
> > > > 	if err != nil {
> > > > 		return nil, err
> > > > 	}
> > > > 	useEarly = true
> > > > case restls.Mode:
> > > > 	var err error
> > > > 	c, err = restls.NewRestls(ctx, c, ss.restlsConfig)
> > > > 	if err != nil {
> > > > 		return nil, fmt.Errorf("%s (restls) connect error: %w", ss.addr, err)
> > > > 	}
> > > > 	useEarly = true
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > See the difference? It doesn't return useEarly = true for case "websocket" and by testing on clash.meta client/xray server it doesn't work either.
> > > 
> > > 
> > > 你就不能在往下看一行么
> > > ```go
> > > useEarly = useEarly || N.NeedHandshake(c)
> > > ```
> > 
> > 
> > My bad, I think I figured out the issue, the NeedHandshake function isn't defined in /transport/v2ray-plugin/websocket.go.
> 
> 如果你不能确定问题在哪，甚至是有没有问题请不要浪费大家的时间 `NewV2rayObfs`中的返回值是由`vmess.StreamWebsocketConn`产生的，`StreamWebsocketConn`会在有`earlyData`配置的时候返回`*websocketWithEarlyDataConn`，该类型是实现了`NeedHandshake`的

https://github.com/MetaCubeX/Clash.Meta/pull/497
Fixed.
希望我没有浪费大家的时间。
